File: src/lib/tauri/renderer.ts
import { invoke } from "@tauri-apps/api/core";

/** How the interface is being drawn. */
export type RenderMode = "gpu" | "software";

/** Why that mode was chosen. */
export type RenderReason =
  /** Nothing asked for anything else. */
  | "default"
  /** `WAVEFLOW_RENDERER` named it. */
  | "forced"
  /** The previous launch never reported a paint, so this one fell back. */
  | "previous-launch-never-painted"
  /** A software launch painted, so this one did not try the GPU. */
  | "remembered"
  /** Software did not paint either, so this one is back on the default. */
  | "software-did-not-help"
  /** The fallback was called for and this platform has none. */
  | "software-unavailable";

export interface RendererStatus {
  mode: RenderMode;
  reason: RenderReason;
  /**
   * Whether `rendererRetryGpu` has anything to undo. False when the mode
   * came from the environment variable — the stored state is not what is
   * deciding then, so the button would change nothing.
   */
  canRetryGpu: boolean;
}

/**
 * How this launch is drawing, and why.
 *
 * `null` when the decision never ran (no app-data directory), in which
 * case the UI shows nothing rather than guessing at a mode.
 */
export function rendererStatus(): Promise<RendererStatus | null> {
  return invoke<RendererStatus | null>("renderer_status");
}

/**
 * Forget the software fallback so the next launch tries the GPU again.
 *
 * Takes effect on the **next** start: the web engine read its
 * environment when its process began and nothing can move it now, which
 * is why every caller has to say so rather than imply the change is
 * live.
 */
export function rendererRetryGpu(): Promise<void> {
  return invoke<void>("renderer_retry_gpu");
}

File: src-tauri/crates/app/src/commands/renderer.rs
//! What the interface is told about how it is being drawn (#595).
//!
//! A silent downgrade is its own kind of confusing: someone whose app
//! has quietly stopped using the GPU deserves to know that, and to have
//! a way back that does not involve finding a file. Both live here —
//! the reading, and the undo.
//!
//! The decision itself is made long before any of this, in
//! [`crate::render_mode`], because it has to happen before a window
//! exists.

use crate::error::{AppError, AppResult};
use crate::render_mode::{self, RenderDecision};

/// How this launch is drawing, and why.
///
/// `None` only when the decision never ran — no app-data directory, so
/// nothing could be armed or remembered. The interface shows nothing
/// rather than guessing at a mode it cannot know.
#[tauri::command]
pub fn renderer_status() -> Option<RenderDecision> {
    render_mode::current()
}

/// Forget the software fallback, so the next launch tries the GPU.
///
/// Takes effect on the next start and says nothing about this one:
/// the web engine read its environment when its process began, and
/// nothing can move it now. The interface is what tells the user that.
#[tauri::command]
pub fn renderer_retry_gpu() -> AppResult<()> {
    let Some(decision) = render_mode::current() else {
        return Err(AppError::Other(
            "the renderer state is not available this launch".into(),
        ));
    };
    if !decision.can_retry_gpu {
        // Nothing to undo. Refused rather than pretended, so a button
        // that cannot help does not report that it did.
        return Err(AppError::Other(
            "nothing to undo: this launch's renderer was not chosen from the stored state".into(),
        ));
    }
    render_mode::retry_gpu();
    Ok(())
}
